<h1> Soal 1: Pemahaman Tentang Model Evaluasi</h1>

Jawab pertanyaan di bawah ini dengan bahasa masing-masing?

1. Apa perbedaan antara data latih, data validasi, dan data test?
2. Bagaimana cara kita menilai performa suatu model?
3. Apa itu Confusion Matrix? Jelaskan secara lengkap!
4. Apa itu Classification Report dari sklearn?

*Jawab*:
1. <p>Data latih merupakan data yang sudah ada yang digunakan oleh model yang kita bentuk untuk belajar. <p> Data Validasi adalah data yang sudah ada yang digunakan untuk memvalidasi model terhadap data yang telah diprediksi. (Dengan kata lain kita mempunyai data label dari data feature yang bersangkutan, tetapi data feature tersebut tidak digunakan oleh model untuk training). <p> Data Test merupakan data baru yang digunakan untuk diprediksi data labelnya menggunakan model

2. <p> a. Menggunakan metrics accuracy_score (berpotensi gagal jika proporsi ouput data tidak seimbang).  <p>b. Menggunakan metrics Cross Validation <p> c.Menggunakan analisa nilai precision, recall, dan f1-score menggunakan confusion matrix ataupun classification report

3. Confusion matrix didefenisikan sebagai klasifikasi statistik yang memvisualisasiklan nilai hasil prediksi yang dibandingkan dengan data sebenarnya. Nilai tersebut ada 2, yaitu positif dan negatif. Berikut keterangan kolom data dari confusion matrix :
<p>- True Positive, apabila tebakannya bernilai positif dan data aktualnya bernilai positif
<p>- True Negative, apabila tebakannya bernilai negatif dan data aktualnya bernilai negatif
<p>- False Positive, apabila tebakannya bernilai positif dan data aktualnya bernilai negatif
<p>- False Negative, apabila tebakannya bernilai negatif dan data aktualnya bernilai positif
Pada Confusion matrix, ada 2 jenis error :
<p>- Type 1 error = Tebakan salah, bernilai positif
<p>- Type 2 error = Tebakan salah, bernilai negatif

4. Classification_report merupakan fungsi yang digunakan untuk menghasilkan report mengenai klasifikasi metrics, berdasarkan data test (yang kita punya) dengan data prediksi. Perhitungan pada precision, recall, serta f1-score pada fungsi ini didasari pada confusion matrix  

---

<h1>Soal 2: Aplikasi Model Evaluasi</h1>

Kali ini kita akan menggunakan data untuk memprediksi kelangsungan hidup pasien yang telah mengalami operasi payudara. Dengan informasi yang dimiliki terkait pasien, kita akan membuat model untuk memprediksi apakah pasien akan bertahan hidup dalam waktu lebih dari 5 tahun atau tidak.

Lebih Lengkapnya kalian bisa membaca informasi tentang dataset di link berikut: https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.names

Buat model Klasifikasi (Model/Algoritma Bebas) untuk memprediksi status pasien dengan ketentuan sebagai berikut:

1. Bagi kedua data ini menjadi data training dan data test dengan test_size=0.25.
3. Pelajar tentang metrics roc_auc_score kemudian buatlah model dan evaluasi dengan menggunakan teknik cross-validation dengan scoring 'roc_auc'. Baca https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html untuk menggunakan metric roc_auc saat cross-validation.
3. Berapa score rata2 dari model dengan teknik cross-validation tersebut?
4. Prediksi data test dengan model yang telah kalian buat!
5. Bagaimana hasil confusion matrix dari hasil prediksi tersebut?
6. Bagaimana classification report dari hasil prediksi tersebut?
5. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status positive?
6. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status negatif?

In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [ ]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [ ]:
df['survival_status'].value_counts()

1    225
2     81
Name: survival_status, dtype: int64

In [ ]:
# Code here
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score

X = df.drop('survival_status', axis=1)
y = df['survival_status']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=21, stratify=y)

model = KNeighborsClassifier(n_neighbors=7, weights='distance')
cv_score = cross_val_score(model, X, y, scoring='roc_auc', cv=10)

print('Score rata2 dari model dengan teknik cross-validation : ')
print(cv_score.mean())

model.fit(X_train,y_train)
y_predict = model.predict(X_test)

from sklearn.metrics import confusion_matrix, classification_report

# survived_status = 1 => Hidup lebih dari 5 tahun (pos)
# survived_status = 2 => Hidup kurang dari 5 tahun (neg)

print('\n')

print('Confusion Matrix :')
conf_mat = confusion_matrix(y_test, y_predict, labels=[1, 2])
print(conf_mat)

print('\n')

print('Hidup Lebih dari 5 Tahun (1) :')
prec1 = conf_mat[0][0]/(conf_mat[0][0] + conf_mat[1][0])
rec1 = conf_mat[0][0]/(conf_mat[0][0] + conf_mat[0][1])
f1s1 = 2*prec1*rec1/(prec1+rec1)
print(f' Precision : {prec1}')
print(f' Recall : {rec1}')
print(f' f1 Score : {f1s1}')

print('\n')

print('Hidup Kurang dari 5 Tahun (2) :')
prec2 = conf_mat[1][1]/(conf_mat[0][1] + conf_mat[1][1])
rec2 = conf_mat[1][1]/(conf_mat[1][0] + conf_mat[1][1])
f1s2 = 2*prec2*rec2/(prec2+rec2)
print(f' Precision : {prec2}')
print(f' Recall : {rec2}')
print(f' f1 Score : {f1s2}')

print('\n')

print('Classification Report :')
class_report = classification_report(y_test, y_predict)
print(class_report)


Score rata2 dari model dengan teknik cross-validation : 
0.6661410298638559


Confusion Matrix :
[[46 11]
 [14  6]]


Hidup Lebih dari 5 Tahun (1) :
 Precision : 0.7666666666666667
 Recall : 0.8070175438596491
 f1 Score : 0.7863247863247863


Hidup Kurang dari 5 Tahun (2) :
 Precision : 0.35294117647058826
 Recall : 0.3
 f1 Score : 0.3243243243243243


Classification Report :
              precision    recall  f1-score   support

           1       0.77      0.81      0.79        57
           2       0.35      0.30      0.32        20

    accuracy                           0.68        77
   macro avg       0.56      0.55      0.56        77
weighted avg       0.66      0.68      0.67        77



Bedasarkan hasil tersebut maka, dapat dilihat bahwa model ini kurang baik untuk memprediksi pasien yang positif hidup lebih dari 5 tahun (bernilai 1), dimana nilai precisionnya sebesar 0.77, recallnya sebesar 0.81, serta f1-scorenya sebesar 0.79. Model ini sangat buruk untuk memprediksi untuk memprediksi pasien yang negatif untuk hidup lebih dari 5 tahun (bernilai 2), dimana nilai precision sebesar 0.35, recallnya sebesar 0.30, serta f1-score sebesar 0.32. Secara keseluruhan, model ini buruk karena memiliki akurasi sebesar 66.614% (scoring 'roc_auc')

---

<h1> Soal 3: Pemahaman Tentang Model Selection</h1>

Jelaskan dengan bahasa sendiri!

1. Apa itu Bias dan Variance?
2. Apa itu Overfitting dan Underfitting?
3. Apa yang bisa kita lakukan untuk mengatur kompleksitas dari model?
4. Bagaimana model yang baik?
5. Kapan kita menggunakan GridSearchcv dan kapan menggunakan RandomizedSearchCV?

Jawab:
1. <p>- Bias merupakan error reducable yang didefenisikan sebagai perbedaan dari rata-rata prediksi oleh model yang dibuat terhadap nilai yang sebenarnya, disebabkan oleh model yang terlalu sederhana.
<p>- Variance merupakan error reducable yang didefenisikan sebagai variabilitas dari model prediksi untuk data yang diberikan (ukuran penyebaran data), disebabkan oleh training yang terlalu banyak

2. <p>- Overfitting merupakan fit (penyesuaian) yang berlebihan, sehingga model seakan-akan menghapal data tanpa mengetahui pola trendnya. Overfitting akan mengakibatkan skor training yang baik, tetapi skor validasinya buruk. <p>- Underfitting merupakan fit(penyesuaian) yang kurang, sehingga model tidak dapat menganalisis data secara menyeluruh. Underfitting akan mengakibatkan skor training dan validasi yang buruk.

3. Melakukan tuning hyperparamater, baik itu secara manual ataupun menggunakan fungsi bantuan.

4. <p>- Model yang baik memiliki nilai training scores dan test scores yang tinggi, serta rentang antara keduanya tidak berjauhan. <p>- Model yang baik juga memiliki bias dan variance yang kecil (optimal pada titik cut off/trade off)

5. <p>- GridSearchcv digunakan apabila kita ingin melakukan tuning parameter secara otomatis dengan memperhatikan keseluruhan data trainingnya. Metode ini akan memakan waktu yang lebih lama, tapi memberikan hasil yang lebih representatif dan paling baik. <p>- RandomizedSearchcv digunakan apabila kita ingin melakukan tuning parameter secara otomatis dengan beberapa data training saja yand dipilih secara random. Metode ini akan memakan waktu yang lebih cepat, tapi hasilnya belum tentu yang terbaik dan representatif.

---

<h1> Soal 4: Aplikasi Model Selection</h1>

1. Bagi kedua data berikut ini menjadi data training dan data test dengan test_size=0.25.
2. Gunakan algoritma KNN sebagai model classifier.
3. Gunakan fungsi GridSearchCV untuk hyperparameter tuning dan model selection.
4. jumlah fold bebas!, gunakan scoring 'roc_auc'
5. Definisikan kombinasi hyperparameter untuk model selection dengan GridSearchCV. kombinasi Hyperparameter bebas, baca lagi dokumentasi KNN di link berikut https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html untuk memahami lagi jenis2 hyperparameter di algorithma KNN.
6. Latih model terhadap data training.
7. Apa hyperparameter terbaik untuk kombinasi hyperparameter kalian?
8. Berapa score validasi terbaik dari model tersebut?
9. Prediksi probabilitasi output dari model yang telah di buat terhadap data test. note : gunakan method .predict_proba() untuk menghasilkan output probabilitas
10. Perhatikan bahwa hasil prediksi ada 2, dimana masing2 adalah nilai probabilitas untuk setiap class label. Ambil nilai probabilitas pasien phositive meninggal dalam waktu kurang dari 5 tahun. note : gunakan bantuan attirubte .classes_ untuk mengetahui urutan label dari hasil prediksi probabilitas.
11. Berapa nilai score roc_auc untuk data test?
12. Apakah model anda termasuk baik, overtting, atau underfitting?
13. Ulangi tahap di atas namun kali ini menggunakan algoritma DecisionTreeClassifier dan kalian bisa menggunakan RandomizedSearchCV apabila process training lama. pelajari algoritma DecisionTreeClassifier di linkberikut: https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html?highlight=decisiontreeclassifier#sklearn.tree.DecisionTreeClassifier
14. Bandingkan scorenya dengan Algoritma KNN, mana yang lebih baik?

Note : Data Science adalah experiment, sangat di dimungkinkan memerlukan beberapa kali percobaan untuk mendapatkan hasil yang terbaik! Happy Coding :)

In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [ ]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [ ]:
# Code here

from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.metrics import roc_auc_score
from sklearn.tree import DecisionTreeClassifier
import numpy as np

X = df.drop('survival_status', axis=1)
y = df['survival_status']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=21, stratify=y)

print('Cara KNeighborsClassifier : \n')
model_knn = KNeighborsClassifier()
param_grid_knn = {'n_neighbors':np.arange(5,60), 'weights':['distance','uniform']}
gscv_knn = GridSearchCV(model_knn, param_grid=param_grid_knn, scoring='roc_auc', cv=10)
gscv_knn.fit(X_train,y_train)

print('Hyperparameter terbaik :')
print(gscv_knn.best_params_)

print(f'Score Validasi : {gscv_knn.best_score_}')

out_pred_prob_knn = gscv_knn.predict_proba(X_test)
out_pred_prob_knn_2 = out_pred_prob_knn[:,1]
print(f'Score roc_auc : {roc_auc_score(y_test, out_pred_prob_knn_2)}')

print('\n')
print('############################################################################')
print('\n')

print('Cara DecisionTreeClassifier : \n')
model_dct = DecisionTreeClassifier()
param_grid_dct = {'max_depth':np.arange(2,11), 'min_samples_leaf':np.arange(1,6)}
gscv_dct = GridSearchCV(model_dct, param_grid=param_grid_dct, scoring='roc_auc', cv=10)
gscv_dct.fit(X_train,y_train)

print('Hyperparameter terbaik :')
print(gscv_dct.best_params_)

print(f'Score Validasi : {gscv_dct.best_score_}')

out_pred_prob_dct = gscv_dct.predict_proba(X_test)
out_pred_prob_dct_2 = out_pred_prob_dct[:,1]
print(f'Score roc_auc : {roc_auc_score(y_test, out_pred_prob_dct_2)}')
####



Cara KNeighborsClassifier : 

Hyperparameter terbaik :
{'n_neighbors': 8, 'weights': 'distance'}
Score Validasi : 0.7094012605042017
Score roc_auc : 0.6166666666666667


############################################################################


Cara DecisionTreeClassifier : 

Hyperparameter terbaik :
{'max_depth': 3, 'min_samples_leaf': 1}
Score Validasi : 0.6520220588235295
Score roc_auc : 0.6052631578947368


Pada kasus ini, dapat dilihat berdasarkan data yang diperoleh, bahwa kedua model bersifat underfitting (apabia validasi 0.7 dapat diterima, maka model kNN termasuk overvitting, dikarenakan score test (roc_auc)nya yang buruk serta memiliki rentang yang jauh. Jika dilihat berdasarkan score, maka algoritma KNN lebih baik dibandingkan DecisionTreeRegressor.